pkg/utils/workloadidentity/secret.go

```diff
@@ -0,0 +1,190 @@
+// for workload identities will pick it up and keep a valid workload identity token stored in it.
+type Secret struct {
+	secret *corev1.Secret
+
+	labels      map[string]string
+	annotations map[string]string
+
+	workloadIdentityName           string
+	workloadIdentityNamespace      string
+	workloadIdentityProviderType   string
+	workloadIdentityContextObject  []byte
+	workloadIdentityProviderConfig []byte
+}
+
+// NewSecret creates a new workload identity secret that will be recognized
+// by the token requestor controller for workload identities which will keep
+// a valid workload identity token stored in it.
+func NewSecret(name, namespace string, opts ...SecretOption) (*Secret, error) {
+	workloadIdentitySecret := &Secret{
+		secret: &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+			Data: make(map[string][]byte),
+		},
+	}
+
+	var err error
+	for _, o := range opts {
+		if e := o(workloadIdentitySecret); err != nil {
+			err = errors.Join(err, e)
+		}
+	}
+
+	if len(workloadIdentitySecret.workloadIdentityName) == 0 {
+		err = errors.Join(err, errors.New("workload identity name is not set"))
+	}
+
+	if len(workloadIdentitySecret.workloadIdentityNamespace) == 0 {
+		err = errors.Join(err, errors.New("workload identity namespace is not set"))
+	}
+
+	if len(workloadIdentitySecret.workloadIdentityProviderType) == 0 {
+		err = errors.Join(err, errors.New("workload identity provider type is not set"))
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return workloadIdentitySecret, nil
+}
+
+// WithLabels is an option that can be used to set additional labels to the workload identity secret
+// which are not necessarily correlated with workload identity specific logic.
+func WithLabels(labels map[string]string) SecretOption {
+	return func(s *Secret) error {
+		s.labels = maps.Clone(labels)
+		return nil
+	}
+}
+
+// WithAnnotations is an option that can be used to set additional annotations to the workload identity secret
+// which are not necessarily correlated with workload identity specific logic.
+func WithAnnotations(annotations map[string]string) SecretOption {
+	return func(s *Secret) error {
+		s.annotations = maps.Clone(annotations)
+		return nil
+	}
+}
+
+// ForWorkloadIdentity is an option that correlates the workload identity secret with a specific workload identity.
+// This option is required upon creation of such secret.
+func ForWorkloadIdentity(name, namespace, providerType string) SecretOption {
+	return func(s *Secret) error {
+		s.workloadIdentityName = name
+		s.workloadIdentityNamespace = namespace
+		s.workloadIdentityProviderType = providerType
+		return nil
+	}
+}
+
+// WithWorkloadIdentityProviderConfig is an option that can be used to store
+// provider specific information in the workload identity secret.
+func WithWorkloadIdentityProviderConfig(providerConfig *runtime.RawExtension) SecretOption {
```

Comment:
Maybe just `WithProviderConfig`?

Reply:
Although the length of the function name is not ideal I think it is still easier to grasp for a user of the package what this function will do if we leave the `WorkloadIdentity` in the function (option) name. When importing the package there can be some duplication in the form of `workloadidentity.WithWorkloadIdentityProviderConfig`, but I think that this is acceptable. This is also my opinion for `WithWorkloadIdentityContextObject` and `ForWorkloadIdentity` functions.

Reply:
I don't agree - almost every usage of these functions will be in other packages, so it always looks like your example `workloadidentity.<function>`. The package name already scopes semantically what these functions are related to. It's even an official code style best practice in Go to avoid stuttering (it's just doesn't kick in here because of the `With` or `For` prefixes).

This is an example complaint (when no prefixes are used):

```
file.go:4:6: type name will be used as file.FileName by other packages, and that stutters; consider calling this Name
file.go:6:6: type name will be used as file.FilePath by other packages, and that stutters; consider calling this Path
```